File: raj_customizations/raj_customizations/report/daily_attendance_report/daily_attendance_report.js
frappe.query_reports["Daily Attendance Report"] = {
  filters: [
    {
      fieldname: "from_date",
      label: "Date",
      fieldtype: "Date",
      reqd: 1,
      default: frappe.datetime.get_today(),
    },
  ],

  // Color full row red if first check-in > 09:00
  formatter: function (value, row, column, data, default_formatter) {
    const formatted = default_formatter(value, row, column, data);

    // Only when we have the full row (data) and a datetime present
    if (data && data.check_in_time) {
      try {
        // check_in_time comes as ISO-like string; moment is bundled in desk
        const t = moment(data.check_in_time);
        // Compare to 09:00 of the same day (local time in browser)
        const threshold = moment(t).hour(9).minute(0).second(0);

        // mark the entire row if late
        if (t.isAfter(threshold)) {
          // Find the datatable row and decorate once
          const $row = $(
            `.dt-row[data-row-index="${row}"], .dt-row-odd[data-row-index="${row}"], .dt-row-even[data-row-index="${row}"]`
          );
          if ($row && !$row.hasClass("late-row")) {
            $row.addClass("late-row");
          }
        }
      } catch (e) {
        // ignore formatting errors silently
      }
    }
    return formatted;
  },

  onload: function (report) {
    // Inject a simple CSS for "late" rows (soft red bg + darker red text)
    const styleId = "daily-checkins-late-style";
    if (!document.getElementById(styleId)) {
      const style = document.createElement("style");
      style.id = styleId;
      style.textContent = `
        .late-row .dt-cell__content { color: #9b1c1c !important; }
        .late-row { background: #ffe5e5 !important; }
      `;
      document.head.appendChild(style);
    }
  },
};

File: raj_customizations/raj_customizations/report/daily_attendance_report/daily_attendance_report.py
import frappe
from frappe.utils import getdate, today

def execute(filters=None):
    filters = filters or {}
    date_str = filters.get("from_date") or today()
    the_date = getdate(date_str)

    columns = [
        {"label": "Employee", "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 120},
        {"label": "Employee Name", "fieldname": "employee_name", "fieldtype": "Data", "width": 180},
        {"label": "Shift", "fieldname": "shift", "fieldtype": "Link", "options": "Shift Type", "width": 150},
        {"label": "Check-in Time", "fieldname": "check_in_time", "fieldtype": "Datetime", "width": 170},
    ]

    # first IN on that date; shift = ec.shift OR employee.default_shift
    data = frappe.db.sql(
        """
        SELECT
            ec.employee,
            e.employee_name,
            COALESCE(MIN(ec.shift), e.default_shift) AS shift,
            MIN(ec.time) AS check_in_time
        FROM `tabEmployee Checkin` ec
        LEFT JOIN `tabEmployee` e ON e.name = ec.employee
        WHERE DATE(ec.time) = %(date)s
        GROUP BY ec.employee, e.employee_name, e.default_shift
        ORDER BY MIN(ec.time) ASC
        """,
        {"date": the_date},
        as_dict=True,
    )

    return columns, data
